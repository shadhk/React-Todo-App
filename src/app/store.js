import { configureStore, combineReducers } from "@reduxjs/toolkit"
import todoReducer from "../slices/todoSlice"

const rootReducer = combineReducers({
  todo: todoReducer
})
export const setupStore = preloadedState => {
  return configureStore({
    reducer: rootReducer,
    preloadedState
  })
}
