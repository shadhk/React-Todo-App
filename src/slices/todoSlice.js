import { createSlice } from "@reduxjs/toolkit"

const getInitialTodo = () => {
  //getting todo list
  const localTodoList = window.localStorage.getItem("todoList")
  //if todo list is not empty
  if (localTodoList) {
    return JSON.parse(localTodoList)
  }
  window.localStorage.setItem("todoList", [])
  return []
}

const initialValue = {
  filterStatus: "all",
  todoList: getInitialTodo()
}

export const todoSlice = createSlice({
  name: "todo",
  initialState: initialValue,
  reducers: {
    addTodo: (state, { payload }) => {
      state.todoList.push(payload)
      const todoList = window.localStorage.getItem("todoList")
      if (todoList) {
        const todoListArr = JSON.parse(todoList)
        todoListArr.push({
          ...payload
        })
        window.localStorage.setItem("todoList", JSON.stringify(todoListArr))
      } else {
        window.localStorage.setItem(
          "todoList",
          JSON.stringify([
            {
              ...payload
            }
          ])
        )
      }
    },
    updateTodo: (state, { payload }) => {
      const todoList = window.localStorage.getItem("todoList")
      if (todoList) {
        const todoListArr = JSON.parse(todoList)
        todoListArr.forEach(todo => {
          if (todo.id === payload.id) {
            todo.status = payload.status
            todo.title = payload.title
          }
        })
        window.localStorage.setItem("todoList", JSON.stringify(todoListArr))
        state.todoList = [...todoListArr]
      }
    },
    deleteTodo: (state, { payload }) => {
      const todoList = window.localStorage.getItem("todoList")
      if (todoList) {
        const todoListArr = JSON.parse(todoList)
        todoListArr.forEach((todo, index) => {
          if (todo.id === payload) {
            todoListArr.splice(index, 1)
          }
        })
        window.localStorage.setItem("todoList", JSON.stringify(todoListArr))
        state.todoList = todoListArr
      }
    },
    updateFilterStatus: (state, { payload }) => {
      state.filterStatus = payload
    }
  }
})

export const { addTodo, updateTodo, deleteTodo, updateFilterStatus } = todoSlice.actions
export default todoSlice.reducer
