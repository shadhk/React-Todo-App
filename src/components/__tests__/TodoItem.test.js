import React from "react"
import { screen } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"
import { describe, expect, it } from "@jest/globals"
import { renderWithProviders } from "../utils/test-utils"
import TodoItem from "../TodoItem"
import { setupStore } from "../../app/store"
import { addTodo } from "../../slices/todoSlice"

describe("React Todo App", () => {
  it("Checking that todo item matches snapshot", async () => {
    const div = document.createElement("div")
    const view = renderWithProviders(<TodoItem />, div)
    expect(view.container).toMatchSnapshot()
  })

  it("render input of add task", async () => {
    const store = setupStore()
    store.dispatch(addTodo({ id: 0, title: "TodoApp", stutus: "complete" }))
    renderWithProviders(<TodoItem />, { store })
    const findid = screen.queryByTestId("todoItem")
    expect(findid).toBeDefined()
  })

  it("render checkButton of add content", async () => {
    renderWithProviders(<TodoItem />)
    const check = screen.queryByTestId("checkbox")
    expect(check).toBeDefined()
  })
})
