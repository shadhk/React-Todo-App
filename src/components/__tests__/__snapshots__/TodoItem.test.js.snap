// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Todo App Checking that todo item matches snapshot 1`] = `
<div>
  <div
    class="item"
    data-testid="todoItem"
  >
    <div
      class="todoDetails"
    >
      <div
        class="svgBox"
        data-testid="button-animate"
      >
        <svg
          class="svg"
          fill="none"
          viewBox="0 0 53 38"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.5 22L16 36.5L51.5 1"
            fill="none"
            opacity="0"
            pathLength="1"
            stroke-dasharray="0px 1px"
            stroke-dashoffset="0px"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="10"
            stroke-width="6"
          />
        </svg>
      </div>
      <div
        class="texts"
      >
        <p
          class="todoText false"
        />
        <p
          class="time"
        >
          2:58 PM, 11/10/2023
        </p>
      </div>
    </div>
    <div
      class="todoActions"
    >
      <div
        class="icon"
        role="button"
        tabindex="0"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
      <div
        class="icon"
        role="button"
        tabindex="0"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
